feat: nodes denylist

psychedelicious opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission


## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

## Description

[feat(backend): allow/deny nodes](https://github.com/invoke-ai/InvokeAI/commit/076a9fa9f0de5e17cbbdb44b807aeeec65338e12)

Allow denying and explicitly allowing nodes. When a not-allowed node is used, a pydantic `ValidationError` will be raised.

- When collecting all invocations, check against the allowlist and denylist first. When pydantic constructs any unions related to nodes, the denied nodes will be omitted
- Add `allow_nodes` and `deny_nodes` to `InvokeAIAppConfig`. These are `Union[list[str], None]`, and may be populated with the `type` of invocations.
- When `allow_nodes` is `None`, allow all nodes, else if it is `list[str]`, only allow nodes in the list
- When `deny_nodes` is `None`, deny no nodes, else if it is `list[str]`, deny nodes in the list
- `deny_nodes` overrides `allow_nodes`

[feat(ui): add UI-level nodes denylist](https://github.com/invoke-ai/InvokeAI/commit/2dbe4a0d6a6a6dbc67f0e6616a35b83891875209)

This simply hides nodes from the workflow editor. The nodes will still work if an API request is made with them. For example, you could hide `iterate` nodes from the workflow editor, but if the Linear UI makes use of those nodes, they will still function.

- Update `AppConfig` with optional property `nodesDenylist: string[]`
- If provided, nodes are filtered out by `type` in the workflow editor

## QA Instructions, Screenshots, Recordings

### Testing the backend denylist handling

Test the existing behaviour:
1. Open the node editor 
2. Reset it if you have anything in there currently
3. Add a `Float Primitive` node
4. Run the workflow with only this single node to confirm it works

Test allowing/denying nodes:
1. Update your `invokeai.yaml` file, adding a new category for `Nodes` with keys for `allow_nodes` and `deny_nodes`. Suggest using this example stanza:
```yaml
  Nodes:
    allow_nodes:
      - add
      - blank_image
      - boolean
      - boolean_collection
      - canny_image_processor
      - clip_skip
      - collect
      - color
      - color_correct
      - compel
      - conditioning
      - conditioning_collection
      - content_shuffle_image_processor
      - controlnet
      - create_denoise_mask
      - cv_inpaint
      - denoise_latents
      - div
      - dynamic_prompt
      - esrgan
      - float
      - float_collection
      - float_range
      - graph
      - hed_image_processor
      - i2l
      - image
      - image_collection
      - img_blur
      - img_chan
      - img_channel_multiply
      - img_channel_offset
      - img_conv
      - img_crop
      - img_hue_adjust
      - img_ilerp
      - img_lerp
      - img_mul
      - img_nsfw
      - img_paste
      - img_resize
      - img_scale
      - img_watermark
      - infill_cv2
      - infill_lama
      - infill_patchmatch
      - infill_rgba
      - infill_tile
      - integer
      - integer_collection
      - iterate
      - l2i
      - l2i_onnx
      - latents
      - latents_collection
      - lblend
      - leres_image_processor
      - lineart_anime_image_processor
      - lineart_image_processor
      - lora_loader
      - lresize
      - lscale
      - main_model_loader
      - mask_combine
      - mask_edge
      - mediapipe_face_processor
      - metadata_accumulator
      - midas_depth_image_processor
      - mlsd_image_processor
      - mul
      - noise
      - normalbae_image_processor
      - onnx_model_loader
      - openpose_image_processor
      - pidi_image_processor
      - prompt_onnx
      - rand_int
      - random_range
      - range
      - range_of_size
      - scheduler
      - sdxl_compel_prompt
      - sdxl_lora_loader
      - sdxl_model_loader
      - sdxl_refiner_compel_prompt
      - sdxl_refiner_model_loader
      - seamless
      - segment_anything_processor
      - step_param_easing
      - string
      - string_collection
      - sub
      - t2l_onnx
      - tile_image_processor
      - tomask
      - vae_loader
      - zoe_depth_image_processor
    deny_nodes:
      - image_processor
      - prompt_from_file
      - show_image
```

2. Restart the server and refresh the web app
3. You should still be able to execute the workflow with a single `float` node, as we haven't denied it yet
4. But, you should not be able to find `ShowImage` in the add nodes search box, because it is denied

Add node to denylist that is already in-use in a workflow:
1. In `invokeai.yaml`, add `float` to `deny_nodes`
2. Restart the server
3. ! Do not refresh the web app, else you'll need to start over !
4. Click Invoke
5. You should get an error about an invalid discriminant value, demonstrating that the backend doesn't care if the UI still thinks the node is valid

Slightly different case
1. Refresh the web app
2. The existing float node should now show an error.
3. Click the workflow editor settings and disable validation
4. Invoke 
5. You should get the same error about an invalid discriminant value.
6. Delete the `float` node and try to add it again - you shouldn't be able to find it in the list

### Testing the UI-level denylist handling

1. Open `invokeai/frontend/web/src/main.tsx`
2. Add this config with nodes denylist: 
![image](https://github.com/invoke-ai/InvokeAI/assets/4822129/bf613770-cc0b-47c3-b774-ef137b097197)
3. Refresh the web app
4. Open workflow editor and you shouldn't be able to find these nodes
5. Go to Linear UI, set image count to 2 and invoke. It should still work (internally it will use `iterate` nodes, but because this is only a UI-level denylist, the backend will still process everything as you'd expect)

psychedelicious commented:
@ebr @maryhipp

maryhipp commented on invokeai/frontend/web/src/app/types/invokeai.ts:
```diff
@@ -49,6 +49,7 @@ export type AppConfig = {
   disabledFeatures: AppFeature[];
   disabledSDFeatures: SDFeature[];
   canRestoreDeletedImagesFromBin: boolean;
+  nodesDenylist: string[];
```
What do you think about making this an `allowList` instead so that we have to explicitly allow nodes to be in there? Otherwise we will be on the defense if there are nodes we want to exclude

psychedelicious replied:
I'll add an allowlist so the behavior mirrors that of the backend filtering.

ebr commented on invokeai/app/invocations/baseinvocation.py:
```diff
@@ -470,14 +472,32 @@ class BaseInvocation(ABC, BaseModel):
     @classmethod
     def get_all_subclasses(cls):
+        app_config = InvokeAIAppConfig.get_config()
+        app_config.parse_args()
```
```suggestion
```

As per https://github.com/invoke-ai/InvokeAI/pull/3466: "In non-main scripts, it is not necessary (or recommended) to call parse_args()" - it also causes issues with unrecognized CLI args, such as `uvicorn` args used in commercial dev. I verified this is working without `parse_args()`.

psychedelicious replied:
It does not work without parse_args unless parse_args is called before this function is called in graph.py. The alternative is to call parse_args in api_app before and of the api stuff is imported, but this seems easier to break

At any rate, we are calling parse_args in api_app anyways, so I can't imagine it's a major issue?

psychedelicious replied:
Ok, I see how this won't work for commercial.

The problem is when the Graph model is created, we need `InvokeAIAppConfig` to be fully loaded - eg, we need `parse_args()` to have been called once. Then, when the union of valid nodes is constructed, they are filtered out by the config.

When you tested and It *appeared* to work without calling `parse_args()`, it only hid the nodes from the UI. That it hid the nodes was an accidental side-effect related to how the openapi schema is massaged for use by the UI. The backend still happily executes the denied nodes.

If `parse_args()` is not allowed due to commercial's set up (FWIW, it *is* called in OSS `api_app.py`), we need to figure out how to ensure the config file has been read by the time we grab the config inside `baseinvocation.py`.

ebr replied:
to be clear, this *does* work for commercial without `parse_args()`

psychedelicious replied:
Reviewed commercial and understand how it *does* actually work, cool.

I've update the PR to work for OSS without needing to call `parse_args()` in `baseinvocation.py` - by rearranging the imports in `api_app.py` and `cli_app.py`.

ebr commented:
Do we want to update this error toast (e.g. when the `iterate` node is disabled on the backend, but used by the linear UI): 
<img src="https://github.com/invoke-ai/InvokeAI/assets/2433541/688b2744-d36f-4475-ba21-e00512eb5be4" width=400 />

Or should we rely (perhaps just temporarily) on whomever modifying this config to make sane choices?

psychedelicious commented:
@ebr it's unfortunately rather tricky to change this because almost all pydantic errors are the same type - ValidationError. There's no easy way to filter out this specific error because invalid discriminator can happen in many cases, and it is an error you'd want to see. 

Could truncate the error messages when they exceed a limit though.

psychedelicious commented:
I've truncated the error messages to 128 characters, and added `nodesAllowlist` to the UI config. The behaviour of this and the backend allow/deny-lists is identical.

psychedelicious commented:
I think this is ready for merge, with a caveat:
- Resolved the issues with calling `parse_args()` inside `baseinvocation.py`
- Added handling for unknown args for the config, superseding #4216 
- Added test to verify that denied nodes are actually denied. Unfortunately, due to how `pytest` collects test modules before running, the test fails when run with the full app suite. It passes when run in isolation. Fixing this is beyond me.

lstein reviewed (approved):
I don't see any problem with the changes to `base.py`.
